Improve legacy QuantizedLinear functions to reduce overhead

xiaomengy opened the pull request:
Summary:
Improve legacy QuantizedLinear functions to reduce overhead.

The legacy QuantizedLinear functions contains may unnecessary zeros and view ops which brings a large amount of overhead. This PR reduces them and improve the existing model performance. 

Differential Revision: D18494988

jianyuh commented on aten/src/ATen/native/QuantizedLinear.cpp:
```diff
@@ -46,125 +45,120 @@ Tensor fbgemm_linear_int8_weight_fp32_activation(
-        /*scale=*/q_params.scale,
-        /*zero_pt=*/q_params.zero_point);
+    fbgemm::PackAWithQuantRowOffset<uint8_t> pack_a(
+        /* trans = */ fbgemm::matrix_op_t::NoTranspose,
```
Nit: the spaces added in the comment: Is it some CPP standard? Changing this might mess up the blame.

BIT-silence replied:
Done.

jianyuh commented on aten/src/ATen/native/QuantizedLinear.cpp:
```diff
@@ -46,125 +45,120 @@ Tensor fbgemm_linear_int8_weight_fp32_activation(
       static_cast<int32_t>(weight_zero_point.to<int64_t>());
 
-  auto bias_contig = bias.contiguous();
+  const auto& bias_contig = bias.is_contiguous() ? bias : bias.contiguous();
```
Wonder if the `bias.contiguous();` implementation under the hood has got such checks?

BIT-silence replied:
Actually the difference is if we use reference here, there will be no tensor shallow copy at all. But anyway we can change it back since it is not a serious thing.

jianyuh reviewed (approved):
LGTM! Any optimizations applied here can be also applied to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp ? Maybe another PR.

xiaomengy committed:
Improve legacy QuantizedLinear functions to reduce overhead (#29773)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/29773

Improve legacy QuantizedLinear functions to reduce overhead.
Separate from the stack of D18381988.

Test Plan: buck test mode/dev-nosan //caffe2/test:jit -- "quant"

Reviewed By: lly-zero-one

Differential Revision: D18494988

fbshipit-source-id: 081687ea8d1e9c67f1213930a40a482e0090a029

jamesr66a commented on aten/src/ATen/native/QuantizedLinear.cpp:
```diff
@@ -217,30 +212,33 @@ std::tuple<Tensor, Tensor, double, int64_t> fbgemm_linear_quantize_weight(
+  Tensor quantized = at::native::empty_like(
+      weight_contig, weight_contig.options().dtype(at::kChar));
+  // Tensor quantized = at::native::empty_cpu(
+  //     weight_contig.sizes(), weight_contig.options().dtype(at::kChar));
```
This probably shouldn't be here

jamesr66a reviewed (commented):
All the unrelated style changes make it very hard to tell what the actual functional changes from this PR are, and it messes up the blame. Please either refrain from making these changes or separate them out into a separate PR in the future
